azurelinuxagent/ga/remoteaccess.py

```diff
@@ -0,0 +1,158 @@
+#
+
+import datetime
+import glob
+import json
+import operator
+import os
+import os.path
+import pwd
+import random
+import re
+import shutil
+import stat
+import subprocess
+import textwrap
+import time
+import traceback
+import zipfile
+
+import azurelinuxagent.common.conf as conf
+import azurelinuxagent.common.logger as logger
+import azurelinuxagent.common.utils.fileutil as fileutil
+import azurelinuxagent.common.version as version
+import azurelinuxagent.common.protocol.wire
+import azurelinuxagent.common.protocol.metadata as metadata
+
+from datetime import datetime, timedelta
+from pwd import getpwall
+from azurelinuxagent.common.errorstate import ErrorState
+
+from azurelinuxagent.common.event import add_event, WALAEventOperation, elapsed_milliseconds
+from azurelinuxagent.common.exception import ExtensionError, ProtocolError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.protocol.restapi import ExtHandlerStatus, \
+                                                    ExtensionStatus, \
+                                                    ExtensionSubStatus, \
+                                                    VMStatus, ExtHandler, \
+                                                    get_properties, \
+                                                    set_properties
+from azurelinuxagent.common.protocol.metadata import MetadataProtocol
+from azurelinuxagent.common.utils.cryptutil import CryptUtil
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.utils.processutil import capture_from_process
+from azurelinuxagent.common.protocol import get_protocol_util
+from azurelinuxagent.common.version import AGENT_NAME, CURRENT_VERSION
+from azurelinuxagent.common.osutil import get_osutil
+
+REMOTE_USR_EXPIRATION_FORMAT = "%a, %d %b %Y %H:%M:%S %Z"
+DATE_FORMAT = "%Y-%m-%d"
+TRANSPORT_PRIVATE_CERT = "TransportPrivate.pem"
+REMOTE_ACCESS_ACCOUNT_COMMENT = "JIT Account"
+MAX_TRY_ATTEMPT = 5
+FAILED_ATTEMPT_THROTTLE = 1
+
+def get_remote_access_handler():
+    return RemoteAccessHandler()
+
+class RemoteAccessHandler(object):
+    def __init__(self):
+        self.protocol_util = get_protocol_util()
+        self.protocol = None
+        self.os_util = None
+        self.cryptUtil = CryptUtil(conf.get_openssl_cmd())
+        self.remote_access = None
+        self.incarnation = 0
+
+    def run(self):
+        try:
+            if self.os_util is None:
+                self.os_util = get_osutil()
+            if self.os_util.jit_enabled:
+                self.protocol = self.protocol_util.get_protocol()
+                current_incarnation = self.protocol.get_incarnation()
+                if self.incarnation != current_incarnation:
+                    # something changed. Handle remote access if any.
+                    self.incarnation = current_incarnation
+                    self.protocol.client.update_goal_state(True)
```

Comment:
Why are you forcing a goal state update?

Reply:
Because WireClient.get_goal_state just returns what it already has without updating unless what it has is None and WireClient.update_goal_state won't refresh the file because t checks the incarnation file value vs the goal_state incarnation value which will always be teh same... so it just returns without updating.  forced=True seems to be the built in way to update this... I can write a new method, but why?

---
In reply to: [194025778](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194025778) [](ancestors = 194025778)

Reply:
We spoke offline about this - I'd like to get this cache updated in the same manner as the other pieces of goalstate in update_goal_state